### PR TITLE
General functions and iteratively convert session data to NWB

### PR DESCRIPTION
**Pull request by Vigneswaran-Chandrasekaran**
Add General functions for different types of modules and iteratively convert each session data to `NWB` format

### CODE REVIEW

File: +nxpl2nwb/Eye.m [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    eye_area = readNPY(f_eye_area);
+    eye_xy_pos = readNPY(f_eye_xy_pos);
+    % create TimeSeries object for eye area
+    pupil = types.core.TimeSeries('timestamps',  eye_timestamps(:, 2), ...
+                 'data', eye_area', ...
+                 'data_unit', ts_data_unit, ...
+                 'description', ts_description, ...
+                 'comments', ts_comments);
+    % create TimeSeries for eye xy s
+    eye_xy = types.core.TimeSeries('timestamps',eye_timestamps(:, 2), ...
```
Comment (bendichter) [inline +nxpl2nwb/Eye.m:30]:
```suggestion
    eye_xy = types.core.SpatialSeries('timestamps',eye_timestamps(:, 2), ...
```

File: +nxpl2nwb/Eye.m [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                'data', eye_xy_pos, ...
+                'data_unit', xy_data_unit, ...
+                'description', xy_description, ...
+                'comments', xy_comments);
+    %% add both TimeSeries objects to PupilTracking and add to behavior
+    % module
+    pupil_track = types.core.PupilTracking('timeseries', [pupil, eye_xy]);
```
Comment (bendichter) [inline +nxpl2nwb/Eye.m:37]:
```suggestion
    pupil_tracking = types.core.PupilTracking();
    pupil_tracking.set('pupil', pupil);
    eye_tracking = types.core.EyeTracking();
    eye_tracking.set('eye_xy', eye_xy);
```

File: +nxpl2nwb/Eye.m [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    behavior_mod.nwbdatainterface.set(...
+                    'PupilTracking', pupil_track);
```
Comment (bendichter) [inline +nxpl2nwb/Eye.m:39]:
```suggestion
                    'PupilTracking', pupil_tracking);
```

File: +nxpl2nwb/Eye.m [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+end
```
Comment (bendichter) [inline +nxpl2nwb/Eye.m:40]:
```suggestion
    behavior_mod.nwbdatainterface.set(...
                    'EyeTracking', eye_tracking);
end
```

**Comment by Vigneswaran-Chandrasekaran** [comment]
Thanks for the review

Now, the general structure of a function looks something similar to,

```matlab
% Function to convert X data  
function types_core_X = Xconvert(file_prefix, X_dataunit, X_description, X_comments)

    arguments
        file_prefix (1, :) string = ''
        X_dataunit char = 'Unknown'
        X_description char = 'some description goes here'
        X_comments char = 'no comments given'
    end
 
    fname_X_property1 = strcat(file_prefix, 'X.property1.npy');
    fname_X_property1 = strcat(file_prefix, 'X.property2.npy');
    
    % read data files
    X_property1 = readNPY(fname_X_property1);
    X_property2 = readNPY(fname_X_property2);
    
    % create types_core_X object for X
    types_core_X = types.core.types_core_X('X_property1',  X_property1, ...
                                          'X_property2', X_property1, ...
                                          'data_unit', X_dataunit, ...
                                          'description', X_description, ...
                                          'comments', X_comments);
end
```

**Comment by bendichter** [comment]
> Thanks for the review
> 
> Now, the general structure of a function looks something similar to,
> 
> ```matlab
> % Function to convert X data  
> function types_core_X = Xconvert(file_prefix, X_dataunit, X_description, X_comments)
> 
>     arguments
>         file_prefix (1, :) string = ''
>         X_dataunit char = 'Unknown'
>         X_description char = 'some description goes here'
>         X_comments char = 'no comments given'
>     end
>  
>     fname_X_property1 = strcat(file_prefix, 'X.property1.npy');
>     fname_X_property1 = strcat(file_prefix, 'X.property2.npy');
>     
>     % read data files
>     X_property1 = readNPY(fname_X_property1);
>     X_property2 = readNPY(fname_X_property2);
>     
>     % create types_core_X object for X
>     types_core_X = types.core.types_core_X('X_property1',  X_property1, ...
>                                           'X_property2', X_property2, ...
>                                           'data_unit', X_dataunit, ...
>                                           'description', X_description, ...
>                                           'comments', X_comments);
> end
> ```

This looks like a great structure!

File: +nxpl2nwb/ClustersSpikes.m [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+function nwb_file = ClustersSpikes(nwb_file, group_view, ...
+                    file_prefix, descriptions)
+    arguments
+        nwb_file {mustBeA(nwb_file, "NwbFile")}
+        group_view {mustBeA(group_view, "types.untyped.ObjectView")}
+        file_prefix (1, :) string = ''
+    end
+
+    arguments (Repeating)
+        descriptions char
```
Comment (bendichter) [inline +nxpl2nwb/ClustersSpikes.m:11]: @Vigneswaran-Chandrasekaran This way, it is unclear which descriptions go to which data objects. Please change it back throughout.

File: +nxpl2nwb/Eye.m [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                                            spatial_dataunit, spatial_description, ...
+                                            spatial_comments)
+    arguments
+        % prefix of the file / identifier to locate particular session
+        file_prefix (1, :) string = ''
+        pupil_track_dataunit char = 'Unknown'
+        pupil_track_description char = 'Time series denoting Pupil movements'
+        pupil_track_comments char = 'No comments'
+        spatial_dataunit char = 'Unknown'
+        spatial_description char = 'Time series denoting eye xy positions'
```
Comment (bendichter) [inline +nxpl2nwb/Eye.m:14]: This is the right way to do descriptions

File: +nxpl2nwb/LickPiezo.m [hunk 7]
```diff
@@ -0,0 +1,23 @@
+% Function to read Lick piezo data files and add to a TimeSeries object
+function lp_timeseries = LickPiezo(file_prefix, dataunit, description)
+    arguments
+        file_prefix (1, :) string = ''
+        dataunit char = 'Unknown'
+        description char = {'Voltage values from a thin-film piezo '
+                            'connected to the lick spout'}
```
Comment (bendichter) [inline +nxpl2nwb/LickPiezo.m:7]: It looks like this would result in an array description, which is not allowed.

File: +nxpl2nwb/TrialTable.m [hunk 8]
```diff
@@ -0,0 +1,78 @@
+% Function to convert trial data and create TrialTable
+function trials = TrialTable(file_prefix, descriptions)
+    arguments
+        file_prefix (1, :) string = ''
+    end
+    arguments (Repeating)
+        descriptions char
```
Comment (bendichter) [inline +nxpl2nwb/TrialTable.m:7]: Please have individual description args with defaults

File: +nxpl2nwb/ElectrodeTable.m [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                                    probe_location, descriptions)
+
+    arguments
+        nwb_file {mustBeA(nwb_file, "NwbFile")}
+        file_prefix (1, :) string = ''
+        probe_location char = 'Unknown'
+    end
+
+    arguments (Repeating)
+        descriptions char
```
Comment (bendichter) [inline +nxpl2nwb/ElectrodeTable.m:12]: Please use individual description args with defaults

File: +nxpl2nwb/Face.m [hunk 10]
```diff
@@ -0,0 +1,26 @@
+% Convert Face energy behavioral data
+function face_energy = Face(file_prefix, dataunit, description, comments)
+    arguments
+        file_prefix (1, :) string = ''
+        dataunit char = 'Unknown'
+        description char = {'Features extracted from frontal part '
+                            'of the subject'}
```
Comment (bendichter) [inline +nxpl2nwb/Face.m:7]: It looks like this will create a cell array rather than a char

File: +nxpl2nwb/PassiveStim.m [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+function [passive_beep_timeseries, passive_clicks_timseries, ...
+          passive_left, passive_right, passive_white] = PassiveStim(...
+                            file_prefix, data_units, descriptions)
+
+    arguments
+        file_prefix (1, :) string = ''
+    end
+    arguments (Repeating)
+        data_units char
+        descriptions char
```
Comment (bendichter) [inline +nxpl2nwb/PassiveStim.m:12]: Please change these to individual args with defaults

File: +nxpl2nwb/Face.m (hunk 10 above)
Comment (bendichter) [inline +nxpl2nwb/Face.m:7]:
```suggestion
        description char = 'Features extracted from frontal part of the subject'
```

File: +nxpl2nwb/LickPiezo.m (hunk 7 above)
Comment (bendichter) [inline +nxpl2nwb/LickPiezo.m:7]:
```suggestion
        description char = 'Voltage values from a thin-film piezo connected to the lick spout'
```

**Comment by Vigneswaran-Chandrasekaran** [comment]
The output NWB file now looks something like this: [Drive link](https://drive.google.com/file/d/1uSDmrltufwm9UF-ZWejpy9ypVdmNhDN2/view?usp=sharing). The NWB file can be read using both Matnwb and Pynwb.

I think now the PR is ready for your kind review :)

**Comment by bendichter** [comment]
@Vigneswaran-Chandrasekaran the file looks very close!

The biggest problem I see with it is that most of the units have 1 firing rate. Is that right?

**Comment by Vigneswaran-Chandrasekaran** [comment]
I have pushed the changes to fix the problems and the current version of output NWB file looks like this: [Drive link](https://drive.google.com/file/d/19q7KY55J8SKLXCaIx_WFo3MzvN-_K31f/view?usp=sharing)

**Pull request merged**